src/core/geometry/qgsgeometryutils.h

```diff
@@ -823,6 +824,62 @@ class CORE_EXPORT QgsGeometryUtils
      */
     static bool transferFirstMValueToPoint( const QgsPointSequence &points, QgsPoint &point );
 
+    /**
+     * A Z or M dimension is added to \a point if one of the points in the list
+     * \a points contains Z or M value.
+     *
+     * This method is equivalent to successively calling Z and M but avoiding
+     * looping twice over the set of points.
+     *
+     * \param points List of points in which a Z or M point is searched.
+     * \param point The point to update with Z or M dimension and value.
+     * \returns TRUE if the point is updated, FALSE otherwise
+     *
+     * \warning This method does not copy the z or m value of the coordinate from the
+     * points whose z or m value is closest to the original x/y point, but only the first one found.
+     *
+     * \since QGIS 3.20
+     */
+    static bool transferFirstZOrMValueToPoint( const QgsPointSequence &points, QgsPoint &point );
+
+    /**
+     * A Z or M dimension is added to \a point if one of the points in the list
+     * \a points contains Z or M value.
+     *
+     * This method is equivalent to successively calling Z and M but avoiding
+     * looping twice over the set of points.
+     *
+     * \param verticesBegin begin vertex which a Z or M point is searched.
+     * \param verticesEnd end vertex which a Z or M point is searched.
+     * \param point The point to update with Z or M dimension and value.
+     * \returns TRUE if the point is updated, FALSE otherwise
+     *
+     * \warning This method does not copy the z or m value of the coordinate from the
+     * points whose z or m value is closest to the original x/y point, but only the first one found.
+     *
+     * \note Not available in Python bindings
+     * \since QGIS 3.20
+     */
+    static bool transferFirstZOrMValueToPoint( const QgsAbstractGeometry::vertex_iterator &verticesBegin, const QgsAbstractGeometry::vertex_iterator &verticesEnd, QgsPoint &point ) SIP_SKIP;
```

Comment:
Hmm - what I was actually suggesting here was something which worked on a generic iterator of QgsPoint -- so it could work both with QgsAbstractGeometry::vertex_iterator AND QgsPointSequence (and std::vector< QgsPoint >, etc...). It **should** be possible but would require a bit a c++ template work. Up to you if you'd like to go down this route or not, but it's an interesting project if you feel like diving into modern c++ :smile:

Reply:
Sorry, I didn't understand it like that :sweat_smile: 
Of course, I'm not against to explore this route! 
On the other hand, I always thought that we were limiting the use of templates in QGIS -- at least some parts like this. 
For the time being, this method will serve as a replacement for the former method `setZValue` which uses only a 2 or 3 point sequence.

Reply:
done, with convenient methods for QgsGeometry and QgsPointSequence.